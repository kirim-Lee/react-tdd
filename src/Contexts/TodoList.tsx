import { createContext, useEffect, useState } from 'react';

interface IContext {
  readonly todoList: string[];
  readonly addTodo: (toDo: string) => void;
  readonly deleteTodo: (index: number) => void;
}

const TodoListContext = createContext<IContext>({
  todoList: [],
  addTodo: () => {},
  deleteTodo: () => {},
});

interface IProps {
  children: JSX.Element | JSX.Element[];
}

const TODO_LIST = 'TodoList';

const TodoListProvider = ({ children }: IProps): JSX.Element => {
  const [todoList, setTodoList] = useState<string[]>([]);

  const addTodo = (toDo: string) => {
    if (!toDo) return;

    const newList = [...todoList, toDo];
    localStorage.setItem(TODO_LIST, JSON.stringify(newList));
    setTodoList(newList);
  };

  const deleteTodo = (index: number) => {
    let list = [...todoList];
    list.splice(index, 1);
    localStorage.setItem(TODO_LIST, JSON.stringify(list));
    setTodoList(list);
  };

  useEffect(() => {
    const list = localStorage.getItem(TODO_LIST);
    if (list) {
      setTodoList(JSON.parse(list));
    }
  }, []);

  return (
    <TodoListContext.Provider value={{ todoList, addTodo, deleteTodo }}>
      {children}
    </TodoListContext.Provider>
  );
};

export { TodoListContext, TodoListProvider };
