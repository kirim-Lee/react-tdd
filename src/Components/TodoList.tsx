import React, { useContext, useState } from 'react';
import styled from 'styled-components';
import { TodoItem } from './TodoItem';
import { TodoListContext } from '../Contexts/TodoList';

const Container = styled.div`
  min-width: 350px;
  height: 400px;
  overflow-y: scroll;
  border: 1px solid #bdbdbd;
`;

export const TodoList = () => {
  const { todoList, deleteTodo } = useContext(TodoListContext);

  return (
    <Container data-testid="toDoList">
      {todoList.map((item, index) => (
        <TodoItem
          key={item}
          id={index}
          label={item}
          onDelete={() => deleteTodo(index)}
        />
      ))}
    </Container>
  );
};
