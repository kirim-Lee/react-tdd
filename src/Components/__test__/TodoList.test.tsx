import { fireEvent, render, screen } from '@testing-library/react';
import { TodoListProvider } from '../../Contexts/TodoList';
import { TodoList } from '../TodoList';

const TODO_LIST = 'TodoList';

beforeEach(() => {
  localStorage.clear();
});

describe('<TodoList />', () => {
  it('렌더링 테스트', () => {
    const { container } = render(
      <TodoListProvider>
        <TodoList />
      </TodoListProvider>
    );

    const todoList = screen.getByTestId('toDoList');
    expect(todoList).toBeInTheDocument();
    expect(todoList.firstChild).toBeNull();

    expect(container).toMatchSnapshot();
  });

  it('todoList 존재할 경우 노출', () => {
    localStorage.setItem(
      TODO_LIST,
      JSON.stringify(['todo 1', 'todo 2', 'todo 3'])
    );

    render(
      <TodoListProvider>
        <TodoList />
      </TodoListProvider>
    );

    expect(screen.getByText('todo 1')).toBeInTheDocument();
    expect(screen.getByText('todo 2')).toBeInTheDocument();
    expect(screen.getByText('todo 3')).toBeInTheDocument();
    expect(screen.getAllByText('삭제').length).toBe(3);
  });

  it('todoList 삭제 연동 테스트', () => {
    localStorage.setItem(
      TODO_LIST,
      JSON.stringify(['todo 1', 'todo 2', 'todo 3'])
    );

    render(
      <TodoListProvider>
        <TodoList />
      </TodoListProvider>
    );

    const todoItem = screen.getByText('todo 2');
    expect(todoItem).toBeInTheDocument();

    fireEvent.click(todoItem.nextElementSibling);
    expect(todoItem).not.toBeInTheDocument();
    expect(JSON.parse(localStorage.getItem(TODO_LIST))).not.toContain('todo 2');
  });
});
