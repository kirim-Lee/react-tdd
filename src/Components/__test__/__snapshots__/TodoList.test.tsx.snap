// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TodoList /> 렌더링 테스트 1`] = `
<div>
  <div
    class="sc-eDvSVe wpcBq"
    data-testid="toDoList"
  />
</div>
`;
